application_form/services/application.py

```diff
@@ -1,54 +1,78 @@
 import logging
 import uuid
 from datetime import date
+from django.contrib.auth import get_user_model
 from django.db import transaction
 from django.db.models import QuerySet
 from typing import Iterable, List, Optional
 
+from apartment.elastic.queries import get_apartment
 from application_form.enums import (
     ApartmentQueueChangeEventType,
+    ApartmentReservationCancellationReason,
     ApartmentReservationState,
 )
-from application_form.models import Applicant, Application, ApplicationApartment
+from application_form.models import (
+    ApartmentReservation,
+    ApartmentReservationStateChangeEvent,
+    Applicant,
+    Application,
+    ApplicationApartment,
+)
 from application_form.services.queue import (
     add_application_to_queues,
-    remove_application_from_queue,
+    remove_reservation_from_queue,
 )
 from customer.services import get_or_create_customer_from_profiles
 
 _logger = logging.getLogger(__name__)
 
+User = get_user_model()
+
 
-def cancel_haso_application(application_apartment: ApplicationApartment) -> None:
+@transaction.atomic
+def cancel_reservation(
+    apartment_reservation: ApartmentReservation,
+    user: User = None,
+    comment: str = None,
+    cancellation_reason: ApartmentReservationCancellationReason = None,
+) -> ApartmentReservationStateChangeEvent:
     """
-    Mark the application as canceled and remove it from the apartment queue.
+    Mark the reservation as canceled and remove it from the apartment queue.
 
-    If the application has already won the apartment, then the winner for the apartment
+    If the reservation has already won the apartment, then the winner for the apartment
     will be recalculated.
     """
-    was_reserved = application_apartment.apartment_reservation.state in [
-        ApartmentReservationState.RESERVED,
-        ApartmentReservationState.REVIEW,
-    ]
-    apartment_uuid = application_apartment.apartment_uuid
-    remove_application_from_queue(application_apartment)
-    if was_reserved:
-        _reserve_haso_apartment(apartment_uuid)
+    was_reserved = (
+        apartment_reservation.state is not ApartmentReservationState.SUBMITTED
```

Comment:
The states are difficult for sure 🙂. But just excluding one state here makes this easier to understand IMO, because the logic is basically "everything else than submitted is considered reserved". I think I'll just merge this for now, I have a feeling we need to fine-tune this logic anyway in the future 🙂

Reply:
Yeah go ahead, I was just looking for an explanation here and you made it: `everything else than submitted is considered reserved`